docs/source/deep_dive/slowmo_ddp.rst

```diff
@@ -0,0 +1,65 @@
+with 8 nodes is not 8x faster than a run with 1 node as we would expect it to be.
+
+SlowMo Distributed Data Parallel aims to solve this by replacing the exact allreduce between gradients, which is typically done, with an approximate
+averaging of parameters. This approximate averaging reduces both the time spent on communication as well as the synchronization between different
+nodes.  It uses one of the following two algorithms (configurable) as a base algorithm for this purpose -
+
+* `Local <https://arxiv.org/abs/1602.05629>`_ `SGD <https://arxiv.org/abs/1705.09056>`_. This algorithm does an allreduce of the parameters every few iterations.
+
+* `Stochastic Gradient Push <https://arxiv.org/abs/1811.10792>`_ (SGP). This algorithm involves one-to-one communications between nodes.
+
+These base algorithms (LocalSGD and SGP), when used only by themselves, result in reduced accuracy. The `SlowMo <https://arxiv.org/abs/1910.00643>`_
+algorithm removes this accuracy loss by doing a slow momentum step, typically, every 48 iterations.
+
+The training process with SlowMo looks as follows:
+
+1. Compute the forward pass.
+
+2. Compute the backward pass.
+
+3. During the backward pass, using a backward hook, the gradients are synchronized using allreduce across the different GPUs on a node.
+
+4. Perform the ``optimizer.step()`` to update parameters on a node with the gradients of that node.
+
+5. Approximately average the parameters using a base algorithm - one of LocalSGD or SGP (both are described above).
+
+6. Perform the slow momentum update step once every ``slowmo_frequency`` (typically 48) iterations. In this step, the parameters on different
+   nodes are reduced, followed by a ``slowmo_optimizer.step()``. Note that this ``slowmo_optimizer`` is different from the original optimizer,
+   and it is done in a `Zero-1 <https://fairscale.readthedocs.io/en/latest/deep_dive/oss_sdp_fsdp.html>`_ like manner to save memory.
+
+Best practices for using ``SlowMoDistributedDataParallel``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. SlowMo will be useful in deep learning workloads which run on greater than 2 nodes in clusters with a slow interconnect, eg Ethernet.
+
+2. SlowMo should be useful in your workload if the following condition holds:
+
+   :math:`\textrm{time_taken_for_all_reduce_of_gradients} \times (1 - \frac{1}{\textrm{localsgd_frequency}} ) > \textrm{time_taken_for_backward_pass}`
+
+   In clusters with slower interconnect, ``time_taken_for_all_reduce_of_gradients`` will go up, leading to SlowMo being more useful. ``localsgd_frequency``
+   is also an important factor here. More details on varying that to affect performance are in tip 2 of
+   `Performance tips for SlowMoDistributedDataParallel`_.
+
+3. ``slowmo_momentum`` will need to be tuned for obtaining good accuracy. A random search across 4 values from [0.1, 0.2, ..., 0.7] should be good enough
```

Comment:
Should it be 7 values?

Reply:
This is what we usually did earlier. I guess I could also change this to a random search for 4 values between 0.0 and 0.7

Reply:
Oh are you saying you were randomly picking 4 values among `{0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7}` and keeping the best one? That doesn't make much sense IMO, because:
* Assuming this is uniform iid sampling, there's a good chance you'll have duplicates
* If you're unlucky you may miss some of the extremes

If you are going to try only 4 values out of 7, just do a grid search, not a random search. I'm not sure what kind of value you typically ended up with, but maybe a grid search other the (5) values `{0.1, 0.2, 0.3, 0.5, 0.7}` would be a reasonable recommendation?